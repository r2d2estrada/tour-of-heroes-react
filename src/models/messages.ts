export interface Message {
  id: number;
  message: string;
}
