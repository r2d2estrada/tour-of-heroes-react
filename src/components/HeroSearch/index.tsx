import React, { SyntheticEvent, useState } from "react";
import Autocomplete from "react-autocomplete";
import { Link, useHistory } from "react-router-dom";
import { Hero } from "../../models/hero";
import "./HeroSearch.scss";

const HeroSearch: React.FC<{ heroes: Hero[] }> = ({ heroes }) => {
  const history = useHistory();
  const [search, setSearch] = useState("");
  const [searchResult, setSearchResult] = useState(heroes);

  const onChange = (event: SyntheticEvent | any): void => {
    const { value } = event.target;
    setSearch(value);
    setSearchResult(
      heroes.filter(
        (hero) =>
          hero.name.toLocaleLowerCase().indexOf(value.toLocaleLowerCase()) > -1
      )
    );
  };

  const onSelect = (val: any): void => {
    setSearch(val);
    const selectedHero = heroes.find((hero) => hero.name === val);
    history.push(`/heroes/${selectedHero?.id}`);
  };

  return (
    <div className="card hero-search">
      <div className="card-header">
        <h4>HeroSearch</h4>
      </div>
      <div className="card-body">
        <Autocomplete
          getItemValue={(hero: Hero) => hero.name}
          items={searchResult}
          renderItem={(hero: Hero, highlighted: boolean) => (
            <div
              key={hero.id}
              className="hero-search-item"
              style={{ borderBottom: "1px solid lightgray" }}
            >
              <Link
                to={`/heroes/${hero.id}`}
                style={{
                  padding: "1rem",
                  display: "block",
                  textDecoration: "none",
                  background: highlighted ? "#eee" : "transparent",
                }}
              >
                {hero.name}
              </Link>
            </div>
          )}
          value={search}
          onChange={onChange}
          onSelect={onSelect}
          menuStyle={{
            zIndex: 2,
            marginTop: 4,
            border: "1px solid lightgray",
            borderRadius: 2,
            boxShadow: "0 2px 12px rgba(0, 0, 0, 0.1)",
            background: "rgba(255, 255, 255)",
            fontSize: "90%",
            position: "fixed",
            overflow: "auto",
            maxHeight: "50%",
          }}
        />
      </div>
    </div>
  );
};

export default HeroSearch;
