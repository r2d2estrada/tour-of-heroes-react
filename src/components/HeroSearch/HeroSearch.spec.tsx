import React from "react";
import { render } from "@testing-library/react";
import { MemoryRouter as Router } from "react-router-dom";
import HeroSearch from ".";
import { Hero } from "../../models/hero";

describe("HeroSearch", () => {
  let component: any;
  let mockData: Hero[];

  beforeAll(() => {
    mockData = [
      {
        id: 11,
        name: "Dr. Nice",
      },
      {
        id: 12,
        name: "Narco",
      },
      {
        id: 13,
        name: "Mr. Bombastic",
      },
    ];

    component = () => {
      const { container } = render(
        <Router>
          <HeroSearch heroes={mockData} />
        </Router>
      );
      return container;
    };
  });

  it("should match snapshot", () => {
    expect(component()).toMatchSnapshot();
  });
});
