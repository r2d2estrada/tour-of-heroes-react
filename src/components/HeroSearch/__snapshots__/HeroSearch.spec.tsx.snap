// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeroSearch should match snapshot 1`] = `
<div>
  <div
    class="card hero-search"
  >
    <div
      class="card-header"
    >
      <h4>
        HeroSearch
      </h4>
    </div>
    <div
      class="card-body"
    >
      <div
        style="display: inline-block;"
      >
        <input
          aria-autocomplete="list"
          aria-expanded="false"
          autocomplete="off"
          role="combobox"
          value=""
        />
      </div>
    </div>
  </div>
</div>
`;
